icommand.lisp

```diff
@@ -0,0 +1,61 @@
+;; Copyright (C) 2016 Caio Oliveira
+;;
+;;  This file is part of stumpwm.
+;;
+;; stumpwm is free software; you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation; either version 2, or (at your option)
+;; any later version.
+
+;; stumpwm is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this software; see the file COPYING.  If not, see
+;; <http://www.gnu.org/licenses/>.
+
+;; Commentary:
+;;
+;;; Macro for defining interactive command. Just pushes and pops new keymaps.
+;;
+;; Code:
+
+(in-package #:stumpwm)
+
+(export '(icommand-enter-interactive-mode
+          icommand-exit-interactive-mode
+          deficommand))
+
+(defun icommand-enter-interactive-mode (kmap name)
+  "Enters interactive mode" ;; TODO
+  (message (format nil "~s started" name))
+  (push-top-map kmap))
+
+(defun icommand-exit-interactive-mode (name)
+  "Exits imove-focus mode"
+  (message "~s finished" name)
+  (pop-top-map))
+
+(defmacro deficommand (name (&key bootstrap-fn cleanup-fn) &body key-bindings)
+  "Declare an interactive command." ;; TODO
+  (let* ((command (if (listp name) (car name) name))
+         (exit-command (concatenate 'string "EXIT-" (symbol-name command)))
+         (keymap (gensym "m")))
+    `(let ((,keymap (make-sparse-keymap)))
+       ,@(loop for keyb in key-bindings
+            collect `(define-key ,keymap ,@keyb))
+       (define-key ,keymap (kbd "RET") ,exit-command)
+       (define-key ,keymap (kbd "C-g") ,exit-command)
+       (define-key ,keymap (kbd "ESC") ,exit-command)
+       (defcommand ,name () ()
+         ,(concatenate 'string "Starts interactive command \"" (symbol-name command) "\"") ;; TODO Add docstring support
+         ,(let ((start-code `(icommand-enter-interactive-mode ,keymap ,(symbol-name command))))
```

Comment:
you can pass the symbol directly to `icommand-enter-interactive-mdoe`, no need to call symbol-name here